File: app/src/main/java/com/nikola/jakshic/dagger/view/fragment/PeerFragment.java
/*package com.nikola.jakshic.dagger.view.fragment;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.nikola.jakshic.dagger.DaggerApp;
import com.nikola.jakshic.dagger.diffcallback.PeerDiffCallback;
import com.nikola.jakshic.dagger.R;
import com.nikola.jakshic.dagger.util.NetworkUtil;
import com.nikola.jakshic.dagger.view.PeerSortDialog;
import com.nikola.jakshic.dagger.view.adapter.PeerAdapter;
import com.nikola.jakshic.dagger.viewModel.PeerViewModel;
import com.nikola.jakshic.dagger.viewModel.DaggerViewModelFactory;

import javax.inject.Inject;

public class PeerFragment extends Fragment implements PeerSortDialog.OnSortListener {

    @Inject
    DaggerViewModelFactory factory;
    private PeerAdapter mAdapter;
    private PeerViewModel viewModel;
    private long accountId;

    public PeerFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        ((DaggerApp) getActivity().getApplication()).getAppComponent().inject(this);
        super.onAttach(context);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View root = inflater.inflate(R.layout.fragment_peer, container, false);

        accountId = getActivity().getIntent().getLongExtra("account_id", -1);

        viewModel = ViewModelProviders.of(this, factory).get(PeerViewModel.class);
        SwipeRefreshLayout mRefresh = root.findViewById(R.id.swiperefresh_peer);

        mAdapter = new PeerAdapter(getContext(), new PeerDiffCallback());

        RecyclerView recyclerView = root.findViewById(R.id.recview_peer);
        recyclerView.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(mAdapter);
        recyclerView.setHasFixedSize(true);

        View btnSort = root.findViewById(R.id.btn_peer_sort);

        btnSort.setOnClickListener(v -> {
            PeerSortDialog dialog = PeerSortDialog.newInstance();
            dialog.setTargetFragment(this, 200);
            dialog.show(getFragmentManager(), null);
        });

        viewModel.initialFetch(accountId);
        viewModel.getPeers().observe(this, mAdapter::submitList);
        viewModel.getStatus().observe(this, status -> {
            switch (status) {
                case LOADING:
                    mRefresh.setRefreshing(true);
                    break;
                default:
                    mRefresh.setRefreshing(false);
                    break;
            }
        });

        mRefresh.setOnRefreshListener(() -> {
            if (NetworkUtil.isActive(getActivity())) {
                viewModel.fetchPeers(accountId);
            } else {
                Toast.makeText(getActivity(), "Check network connection!", Toast.LENGTH_SHORT).show();
                mRefresh.setRefreshing(false);
            }
        });

        return root;
    }

    @Override
    public void onSort(int sortOption) {

        // Remove previous observers b/c we are attaching new LiveData<PagedList>
        viewModel.getPeers().removeObservers(this);

        switch (sortOption) {
            case 0:
                viewModel.sortByGames(accountId);
                break;
            case 1:
                viewModel.sortByWinrate(accountId);
                break;
        }

        // Set to null first, to delete all the items otherwise the list wont be scrolled to the first item
        mAdapter.submitList(null);

        // Attach the observer to the new LiveData<PagedList>
        viewModel.getPeers().observe(this, mAdapter::submitList);

    }
}*/

File: app/src/main/java/com/nikola/jakshic/dagger/view/fragment/PeerFragment.kt
package com.nikola.jakshic.dagger.view.fragment

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.nikola.jakshic.dagger.*
import com.nikola.jakshic.dagger.diffcallback.PeerDiffCallback
import com.nikola.jakshic.dagger.util.NetworkUtil
import com.nikola.jakshic.dagger.view.PeerSortDialog
import com.nikola.jakshic.dagger.view.adapter.PeerAdapter
import com.nikola.jakshic.dagger.viewModel.DaggerViewModelFactory
import com.nikola.jakshic.dagger.viewModel.PeerViewModel
import kotlinx.android.synthetic.main.fragment_peer.*
import javax.inject.Inject

class PeerFragment : Fragment(), PeerSortDialog.OnSortListener{

    @Inject lateinit var factory: DaggerViewModelFactory
    private var id : Long = -1
    private lateinit var viewModel : PeerViewModel
    private lateinit var adapter: PeerAdapter

    override fun onAttach(context: Context?) {
        (activity?.application as DaggerApp).appComponent.inject(this)
        super.onAttach(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return container?.inflate(R.layout.fragment_peer)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        id = activity?.intent?.getLongExtra("account_id",-1)?: -1

        viewModel = ViewModelProviders.of(this, factory)[PeerViewModel::class.java]

        viewModel.initialFetch(id)

        adapter = PeerAdapter(context, PeerDiffCallback())

        recView.layoutManager = LinearLayoutManager(context)
        recView.addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        recView.adapter = adapter
        recView.setHasFixedSize(true)

        viewModel.peers.observe(this, Observer(adapter::submitList))
        viewModel.status.observe(this, Observer {
            when(it){
                Status.LOADING -> swipeRefresh.isRefreshing = true
                else -> swipeRefresh.isRefreshing = false
            }
        })

        val sortDialog = PeerSortDialog.newInstance()
        sortDialog.setTargetFragment(this, 300)
        btnSort.setOnClickListener {
            sortDialog.show(fragmentManager, null)
        }

        swipeRefresh.setOnRefreshListener {
            if (NetworkUtil.isActive(context))
                viewModel.fetchPeers(id)
            else{
                toast("Check network connection!")
                swipeRefresh.isRefreshing = false
            }
        }
    }

    override fun onSort(sortOption: Int) {
        // Remove previous observers b/c we are attaching new LiveData<PagedList>
        viewModel.peers.removeObservers(this)

        when(sortOption){
            0 -> viewModel.sortByGames(id)
            else -> viewModel.sortByWinrate(id)
        }
        // Set to null first, to delete all the items otherwise the list wont be scrolled to the first item
        adapter.submitList(null)
        // Attach the observer to the new LiveData<PagedList>
        viewModel.peers.observe(this, Observer(adapter::submitList))
    }
}